tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+    }
+    virtual void setIndexBuffer(BufferResource* buffer, Format indexFormat, UInt offset = 0) override
+    {
+        SLANG_UNUSED(buffer);
+        SLANG_UNUSED(indexFormat);
+        SLANG_UNUSED(offset);
+    }
+    virtual void setDepthStencilTarget(ResourceView* depthStencilView) override
+    {
+        SLANG_UNUSED(depthStencilView);
+    }
+    virtual void setViewports(UInt count, Viewport const* viewports) override
+    {
+        SLANG_UNUSED(count);
+        SLANG_UNUSED(viewports);
+    }
+    virtual void setScissorRects(UInt count, ScissorRect const* rects) override
+    {
+        SLANG_UNUSED(count);
+        SLANG_UNUSED(rects);
+    }
+    virtual void draw(UInt vertexCount, UInt startVertex) override
+    {
+        SLANG_UNUSED(vertexCount);
+        SLANG_UNUSED(startVertex);
+    }
+    virtual void drawIndexed(UInt indexCount, UInt startIndex, UInt baseVertex) override
+    {
+        SLANG_UNUSED(indexCount);
+        SLANG_UNUSED(startIndex);
+        SLANG_UNUSED(baseVertex);
+    }
+};
+
+SlangResult CUDAShaderObject::init(Renderer* renderer, CUDAShaderObjectLayout* typeLayout)
+{
+    this->layout = typeLayout;
+    
+    // If the layout tells us that there is any uniform data,
+    // then we need to allocate a constant buffer to hold that data.
+    //
+    // TODO: Do we need to allocate a shadow copy for use from
+    // the CPU?
+    //
+    // TODO: When/where do we bind this constant buffer into
+    // a descriptor set for later use?
+    //
+    auto slangLayout = layout->typeLayout;
+    size_t uniformSize = layout->typeLayout->getSize();
+    if (uniformSize)
+    {
+        BufferResource::Desc bufferDesc;
+        bufferDesc.init(uniformSize);
+        bufferDesc.cpuAccessFlags |= Resource::AccessFlag::Write;
+        RefPtr<BufferResource> constantBuffer;
+        SLANG_RETURN_ON_FAIL(renderer->createBufferResource(
+            Resource::Usage::ConstantBuffer, bufferDesc, nullptr, constantBuffer.writeRef()));
+        bufferResource = dynamic_cast<MemoryCUDAResource*>(constantBuffer.Ptr());
+    }
+
+    // If the layout specifies that we have any sub-objects, then
+    // we need to size the array to account for them.
+    //
+    Index subObjectCount = slangLayout->getSubObjectRangeCount();
+    objects.setCount(subObjectCount);
+
+    for (auto subObjectRange : layout->subObjectRanges)
+    {
+        RefPtr<CUDAShaderObjectLayout> subObjectLayout = subObjectRange.layout;
+
+        // In the case where the sub-object range represents an
+        // existential-type leaf field (e.g., an `IBar`), we
+        // cannot pre-allocate the object(s) to go into that
+        // range, since we can't possibly know what to allocate
+        // at this point.
+        //
+        if (!subObjectLayout)
+            continue;
+        //
+        // Otherwise, we will allocate a sub-object to fill
+        // in each entry in this range, based on the layout
+        // information we already have.
```

Comment:
I'm not 100% certain, but it feels like we'd eventually want to move the automatic allocation of sub-objects into a higher layer that works across all the `ShaderObject` implementations, since it is mostly a utility for applications that don't want to have to think about explicitly managing this stuff.